File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    vector<vector<int>> dp;
    int help(string& t1,string& t2, int i,int j){
        if(i<0 || j<0) return 0;
        if(dp[i][j] != -1) return dp[i][j];
        if(t1[i]==t2[j]) return dp[i][j] = 1 + help(t1,t2,i-1,j-1);
        else return dp[i][j] = max(help(t1,t2,i-1,j),help(t1,t2,i,j-1));
    }
    int longestCommonSubsequence(string text1, string text2) {
    int n = text1.size();
    int m = text2.size();
    dp.resize(n+1,vector<int>(m+1,-1));
    return help(text1,text2,n-1,m-1);    
    }
};

// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
    int n = text1.length();
    int m = text2.length();
    int[][]  dp = new int[n+1][m+1];
    for(int i=1; i<=n; i++){
        for(int j=1; j<=m; j++){
            char ch = text1.charAt(i-1);
            char sh = text2.charAt(j-1);
            if(ch==sh) dp[i][j] = 1 + dp[i-1][j-1];
            else dp[i][j] = Math.max(dp[i][j-1],dp[i-1][j]);
        }
    }  
    return dp[n][m];
    }
}

// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna